File: benchmarks/concurrency/prodcons_10k.py
import time, queue, threading
q = queue.Queue()
result = [0]
def producer():
    for i in range(10000):
        q.put(i)
    q.put(None)
def consumer():
    s = 0
    while True:
        item = q.get()
        if item is None: break
        s += item
    result[0] = s
start = time.monotonic_ns()
p = threading.Thread(target=producer)
c = threading.Thread(target=consumer)
p.start(); c.start()
p.join(); c.join()
end = time.monotonic_ns()
print((end - start) // 1000)

File: benchmarks/concurrency/prodcons_10k.rs
use std::time::Instant;
use std::sync::mpsc;
use std::thread;
fn main() {
    let start = Instant::now();
    let (tx, rx) = mpsc::channel();
    let producer = thread::spawn(move || {
        for i in 0..10000 {
            tx.send(i).unwrap();
        }
    });
    let consumer = thread::spawn(move || {
        let mut sum: i64 = 0;
        for val in rx {
            sum += val;
        }
        sum
    });
    producer.join().unwrap();
    let _ = consumer.join().unwrap();
    println!("{}", start.elapsed().as_micros());
}
